examples/raytracing/example_raytracing.cpp

```diff
@@ -9,81 +9,179 @@
+    auto const &box = _boxes(primitive_index);
+    int const predicate_index = ArborX::getData(predicate);
+    float const kappa = 1.;
+    overlapDistance(ray, box, length, entrylength);
+    float const optical_path_length = kappa * length;
+
+    float const energy_deposited =
+        -_ray_energy(predicate_index) * expm1(-optical_path_length);
+    _ray_energy(predicate_index) += energy_deposited;
+    Kokkos::atomic_add(&_energy(primitive_index), energy_deposited);
+  }
 };
+} // namespace OrderedIntersectsBased
 
 template <typename MemorySpace>
-struct ArborX::AccessTraits<SpheresToBoxes<MemorySpace>, ArborX::PrimitivesTag>
+struct ArborX::AccessTraits<OrderedIntersectsBased::Rays<MemorySpace>,
+                            ArborX::PredicatesTag>
 {
   using memory_space = MemorySpace;
+  using size_type = std::size_t;
 
-  KOKKOS_FUNCTION static std::size_t
-  size(SpheresToBoxes<MemorySpace> const &stob)
+  KOKKOS_FUNCTION
+  static size_type size(OrderedIntersectsBased::Rays<MemorySpace> const &rays)
   {
-    return stob._spheres.extent(0);
+    return rays._rays.extent(0);
   }
-  KOKKOS_FUNCTION static ArborX::Box
-  get(SpheresToBoxes<MemorySpace> const &stobs, std::size_t const i)
+  KOKKOS_FUNCTION
+  static auto get(OrderedIntersectsBased::Rays<MemorySpace> const &rays,
+                  size_type i)
   {
-    auto const &sphere = stobs._spheres(i);
-    auto const &c = sphere.centroid();
-    auto const r = sphere.radius();
-    return {{c[0] - r, c[1] - r, c[2] - r}, {c[0] + r, c[1] + r, c[2] + r}};
+    return attach(ordered_intersects(rays._rays(i)), (int)i);
   }
 };
 
+namespace IntersectsBased
+{
+/*
+ * Storage for the rays and access traits used in the query/traverse.
+ */
 template <typename MemorySpace>
 struct Rays
 {
   Kokkos::View<ArborX::Experimental::Ray *, MemorySpace> _rays;
 };
 
-template <typename MemorySpace>
-struct ArborX::AccessTraits<Rays<MemorySpace>, ArborX::PredicatesTag>
+/*
+ * IntersectedCell is a storage conatiner for all intersection between rays and
+ * boxes that are detected when calling the AccumRaySphereOptDist struct. The
+ * member variables that are relevant for sorting the intersection according to
+ * box and ray are contained in the base class IntersectedCellForSorting as
+ * performance improvement.
+ */
+struct IntersectedCellForSorting
 {
-  using memory_space = MemorySpace;
-
-  KOKKOS_FUNCTION static std::size_t size(Rays<MemorySpace> const &rays)
+  float key;
+  int rid;
+  friend KOKKOS_FUNCTION bool operator<(IntersectedCellForSorting const &l,
+                                        IntersectedCellForSorting const &r)
   {
-    return rays._rays.extent(0);
-  }
-  KOKKOS_FUNCTION static auto get(Rays<MemorySpace> const &rays, std::size_t i)
-  {
-    return attach(intersects(rays._rays(i)), (int)i);
+    if (l.rid == r.rid)
+      return l.key < r.key;
+    return l.rid < r.rid;
   }
 };
 
+struct IntersectedCell : public IntersectedCellForSorting
```

Comment:
If I understand it right, the only reason for inheritance is to save some memory space, so as not to duplicate the partial data. Correct?

Reply:
Yes.